File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

vector<int> solution(string msg) {
    vector<int> answer;
    
    map<string, int> m;
    for(int idx = 0; idx < 26; idx++) {
        string tmp = "";
        tmp += 'A' + idx;
        m.insert({tmp, idx + 1});
    }
    
    int len = msg.length();
    string str = msg.substr(0, 1);
    
    for(int idx = 1; idx < len; idx++) {
        if(m.find(str + msg[idx]) != m.end()) {
            str += msg[idx];
        }     
        else {
            answer.push_back(m[str]);
            m.insert({str + msg[idx], m.size() + 1});
            str = msg.substr(idx, 1);
        }
    }
    answer.push_back(m[str]);    
    
    return answer;
}

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.py
def solution(msg):
    answer = []
    indices = {}
    for i, c in enumerate(range(ord('A'), ord('Z') + 1), start=1):
        indices[chr(c)] = i
        
    total_len = len(msg)
    start = 0
    end = 1
    while True:
        ans_index = -1
        while end <= total_len and msg[start:end] in indices:
            ans_index = indices[msg[start:end]]
            end += 1
            
        answer.append(ans_index)
        
        if end > total_len:
            break
        else:
            indices[msg[start:end]] = len(indices) + 1
            start = end - 1
    
    return answer